drivers/modem/ublox-sara-r4.c

```diff
@@ -0,0 +1,1854 @@
+			   socklen_t addrlen,
+			   net_context_connect_cb_t cb,
+			   s32_t timeout,
+			   void *user_data)
+{
+	int ret;
+	char buf[sizeof("AT+USOCO=#,!###.###.###.###!,#####,#\r")];
+	struct modem_socket *sock;
+
+	if (!context || !addr) {
+		return -EINVAL;
+	}
+
+	sock = (struct modem_socket *)context->offload_context;
+	if (!sock) {
+		LOG_ERR("Can't locate socket for net_ctx:%p!", context);
+		return -EINVAL;
+	}
+
+	if (sock->socket_id < MDM_BASE_SOCKET_NUM - 1) {
+		LOG_ERR("Invalid socket_id(%d) for net_ctx:%p!",
+			    sock->socket_id, context);
+		return -EINVAL;
+	}
+
+	sock->dst.sa_family = addr->sa_family;
+
+#if defined(CONFIG_NET_IPV6)
+	if (addr->sa_family == AF_INET6) {
+		net_ipaddr_copy(&net_sin6(&sock->dst)->sin6_addr,
+				&net_sin6(addr)->sin6_addr);
+		sock->dst_port = ntohs(net_sin6(addr)->sin6_port);
+		net_sin6(&sock->dst)->sin6_port = net_sin6(addr)->sin6_port;
+	} else
+#endif
+#if defined(CONFIG_NET_IPV4)
+	if (addr->sa_family == AF_INET) {
+		net_ipaddr_copy(&net_sin(&sock->dst)->sin_addr,
+				&net_sin(addr)->sin_addr);
+		sock->dst_port = ntohs(net_sin(addr)->sin_port);
+		net_sin(&sock->dst)->sin_port = net_sin6(addr)->sin6_port;
+	} else
+#endif
+	{
+		return -EINVAL;
+	}
+
+	if (sock->dst_port < 0) {
+		LOG_ERR("Invalid port: %d", sock->dst_port);
+		return -EINVAL;
+	}
+
+	/* TODO: timeout is currently ignored */
+
+	/* skip socket connect if UDP */
+	if (sock->ip_proto == IPPROTO_UDP) {
+		return 0;
+	}
+
+	snprintk(buf, sizeof(buf), "AT+USOCO=%d,\"%s\",%d", sock->socket_id,
+		 modem_sprint_ip_addr(addr), sock->dst_port);
+	ret = send_at_cmd(sock, buf, MDM_CMD_CONN_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("%s ret:%d", buf, ret);
+	}
+
+	if (cb) {
+		cb(context, ret, user_data);
+	}
+
+	return ret;
+}
+
+static int offload_accept(struct net_context *context,
+			  net_tcp_accept_cb_t cb,
+			  s32_t timeout,
+			  void *user_data)
+{
+	/* NOT IMPLEMENTED */
+	return -ENOTSUP;
+}
+
+static int offload_sendto(struct net_pkt *pkt,
+			  const struct sockaddr *dst_addr,
+			  socklen_t addrlen,
+			  net_context_send_cb_t cb,
+			  s32_t timeout,
+			  void *user_data)
+{
+	struct net_context *context = net_pkt_context(pkt);
+	struct modem_socket *sock;
+	int ret = 0, dst_port = -1;
+
+	if (!context) {
+		return -EINVAL;
+	}
+
+	sock = (struct modem_socket *)context->offload_context;
+	if (!sock) {
+		LOG_ERR("Can't locate socket for net_ctx:%p!", context);
+		return -EINVAL;
+	}
+
+#if defined(CONFIG_NET_IPV6)
+	if (dst_addr->sa_family == AF_INET6) {
+		dst_port = ntohs(net_sin6(dst_addr)->sin6_port);
+	} else
+#endif
+#if defined(CONFIG_NET_IPV4)
+	if (dst_addr->sa_family == AF_INET) {
+		dst_port = ntohs(net_sin(dst_addr)->sin_port);
+	} else
+#endif
+	{
+		return -EINVAL;
+	}
+
+	if (dst_port < 0) {
+		LOG_ERR("Invalid port: %d", dst_port);
+		return -EINVAL;
+	}
+
+	ret = send_data(sock, dst_addr, dst_port, pkt);
+	if (ret < 0) {
+		LOG_ERR("send_data error: %d", ret);
+	}
+
+	net_pkt_unref(pkt);
```

Comment:
Correct me if I'm wrong, but will not the caller do `net_pkt_unref` here if `ret < 0`? So this should be in an else branch. I'm using DTLS and saw the error `<err> net_pkt: *** ERROR *** pkt 0x20027e78 is freed already by offload_sendto():1717 (context_sendto():1473)`.

Reply:
Also correct here.  Fixing.